cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/purgatory/Purgatory.java

```diff
@@ -0,0 +1,194 @@
+      throw new IllegalArgumentException(String.format("Purgatory can only contain POST request (Attempted to add: %s).",
+                                                       httpServletRequestToString(request)));
+    }
+    RequestInfo requestInfo = new RequestInfo(request, parameters);
+    _requestInfoById.put(_requestId, requestInfo);
+
+    return new PurgatoryOrReviewResult(Collections.singletonMap(_requestId++, requestInfo));
+  }
+
+  /**
+   * Ensure that:
+   * <ul>
+   *   <li>A request with the given review id exists in the purgatory.</li>
+   *   <li>The request with the given review id matches the given request.</li>
+   *   <li>The request with the given review id is approved in the purgatory.</li>
+   * </ul>
+   *
+   * Then mark the review status as submitted.
+   *
+   * @param reviewId The review id for which the corresponding request is requested to be submitted.
+   * @param request The request to submit.
+   * @return Submitted request info.
+   */
+  public synchronized RequestInfo submit(int reviewId, HttpServletRequest request) {
+    RequestInfo requestInfo = _requestInfoById.get(reviewId);
+    // 1. Ensure that a request with the given review id exists in the purgatory.
+    if (requestInfo == null) {
+      throw new UserRequestException(
+          String.format("No request with review id %d exists in purgatory. Please use %s endpoint to check for the "
+                        + "current requests awaiting review in purgatory.", reviewId, REVIEW));
+    }
+
+    // 2. Ensure that the request with the given review id matches the given request.
+    EndPoint endpoint = ParameterUtils.endPoint(request);
+    if (requestInfo.endPoint() != endpoint) {
+      throw new UserRequestException(
+          String.format("Request with review id %d is associated with %s endpoint, but the given request has %s endpoint."
+                        + "Please use %s endpoint to check for the current requests awaiting review in purgatory.",
+                        reviewId, requestInfo.endPoint(), endpoint, REVIEW));
+    }
+
+    if (requestInfo.status() == ReviewStatus.SUBMITTED && LOG.isDebugEnabled()) {
+        LOG.info("Request {} has already been submitted for execution.", requestInfo.endpointWithParams());
+    } else {
+      // 3. Ensure that the request with the given review id is approved in the purgatory, and mark the status as submitted.
+      requestInfo.applyReview(ReviewStatus.SUBMITTED, FINAL_REASON);
+      LOG.info("Submitted request {} for execution.", requestInfo.endpointWithParams());
+    }
+    return requestInfo;
+  }
+
+  /**
+   * Remove the {@link ReviewStatus#SUBMITTED} request associated with the given review id from purgatory.
+   *
+   * @param reviewId Review id of the request to be removed from the purgatory.
+   * @return Removed submitted request if exists in purgatory, null otherwise.
+   */
+  public synchronized RequestInfo removeSubmitted(int reviewId) {
```

Comment:
Do we allow removal of old pending requests? If a request is approved or is sitting to be approved for too long, can we still discard it?

Reply:
Transitions between different `ReviewStatus` is documented in `RequestInfo` JavaDoc. 
```
 * <li>{@link ReviewStatus#PENDING_REVIEW} -> {@link ReviewStatus#APPROVED}, {@link ReviewStatus#DISCARDED}</li>
```

Submitted tasks are removed if a completed user is removed from user task manager to prevent users from calling a new request with a previously approved requestId that has already executed and expired.